### ISSUE TITLE
Add HasCallStack to partial functions

### ISSUE THREAD

**Issue by dwincort** [haskell/containers#489]
Can we add `HasCallStack` to the constraints of partial functions like `Map.!`?  I'd be happy to put together a PR if it's a change that would get accepted.

**Comment by treeowl** [comment]
Yes, please! I was already thinking about doing that. But please show
benchmark results to make sure the performance impact is tolerable.

**Comment by sjakobi** [comment]
Related: https://github.com/haskell-unordered-containers/unordered-containers/pull/252

**Comment by sjakobi** [comment]
Since the analogous work on `unordered-containers` was very warmly received, I'd love to see progress on this issue.

AFAICT not all changes in #493 are problematic. I think it might be a good first step to extract those into a separate PR that can be merged quickly.

(CC @bollu)

**Comment by dwincort** [comment]
Oh wow, I totally forgot about this.  The project I was working on ended up finding a different way to track errors, and this went to the bottom of my priority list.

I don't have the bandwidth right now to work on this, so if someone else wants to take over, that would be great.  Otherwise, I may be able to find some time in a few weeks.  I'll make a reminder for myself.

**Comment by sjakobi** [comment]
For reference: Similar plans for `base` are discussed in https://gitlab.haskell.org/ghc/ghc/-/issues/17040.

**Comment by treeowl** [comment]
Yes, I tend to think it's acceptable if minimal. I think the theoretical justification is that it can help an end-user of software using the library write a useful bug report if said software has a bug.
